core/src/apps/zcash/orchard/signer.py

```diff
@@ -0,0 +1,300 @@
+        tx_info: TxInfo,
+        seed: bytes,
+        approver: ZcashApprover,
+        coin: CoinInfo,
+        tx_req: TxRequest,
+    ) -> None:
+        assert tx_req.serialized is not None  # typing
+
+        self.inputs_count = tx_info.tx.orchard_inputs_count
+        self.outputs_count = tx_info.tx.orchard_outputs_count
+
+        if self.inputs_count + self.outputs_count > 0:
+            self.actions_count = max(
+                2,  # minimal required amount of actions
+                self.inputs_count,
+                self.outputs_count,
+            )
+        else:
+            self.actions_count = 0
+
+        if self.actions_count == 0:
+            return  # no need to initialize other attributes
+
+        self.tx_info = tx_info
+        self.keychain = OrchardKeychain.from_seed_and_coin(seed, coin)
+        self.approver = approver
+        self.coin = coin
+        self.tx_req = tx_req
+        assert isinstance(tx_info.sig_hasher, ZcashHasher)
+        self.sig_hasher: ZcashHasher = tx_info.sig_hasher
+
+        account = tx_info.tx.account
+        assert account is not None  # typing
+        key_path = [
+            32 | HARDENED,  # ZIP-32 constant
+            coin.slip44 | HARDENED,  # purpose
+            account | HARDENED,  # account
+        ]
+        self.key_node = self.keychain.derive(key_path)
+
+        self.msg_acc = MessageAccumulator(
+            self.keychain.derive_slip21(
+                [b"Zcash Orchard", b"Message Accumulator"],
+            ).key()
+        )
+
+        self.rng = None
+
+    @skip_if_empty
+    async def process_inputs(self) -> None:
+        await self.check_orchard_inputs_count()
+        for i in range(self.inputs_count):
+            txi = await self.get_input(i)
+            self.msg_acc.xor_message(txi, i)  # add message to the accumulator
+            self.approver.add_orchard_input(txi)
+
+    def check_orchard_inputs_count(self) -> Awaitable[None]:  # type: ignore [awaitable-is-generator]
+        if self.inputs_count > MAX_SILENT_ORCHARD_INPUTS:
+            yield ConfirmOrchardInputsCountOverThreshold(self.inputs_count)
+
+    @skip_if_empty
+    async def approve_outputs(self) -> None:
+        for i in range(self.outputs_count):
+            txo = await self.get_output(i)
+            self.msg_acc.xor_message(txo, i)  # add message to the accumulator
+            if output_is_internal(txo):
+                self.approver.add_orchard_change_output(txo)
+            else:
+                await self.approver.add_orchard_external_output(txo)
+
+    @skip_if_empty
+    async def compute_digest(self) -> None:
+        # derive shielding seed
+        shielding_seed = self.derive_shielding_seed()
+        self.rng = BundleShieldingRng(seed=shielding_seed)
+
+        # send shielded_seed to the host
+        assert self.tx_req.serialized is not None  # typing
+        self.tx_req.serialized.zcash_shielding_seed = shielding_seed
+        await self.release_serialized()
```

Comment:
we want to emit the zcash_shielding_seed **before** shielding procedure (to parallelize the signing and the proof computation)

a special case can occur for transactions with <2 shielded inputs **and** <2 shielded outputs:

signer pads this tx by dummy input(s) and dummy output(s)

after shuffling , it can happen that the first that first Action consists from pair of a dummy input and dummy output 

it follows that Trezor does not request any data before shielding the first Action and the zcash_shielding_seed is sent later , in the middle of shielding procedure

in fact , this "special case" is quite significant , since it occurs with prob. 50% for shielded-to-transparent (z2t) and transparent-to-shielded (t2z) transactions